dist/ExtUtils-ParseXS/lib/perlxs.pod

```diff
@@ -1,2065 +1,4706 @@
+
+There are similar macros
+
+    XSRETURN_YES
+    XSRETURN_NO
+    XSRETURN_EMPTY
+
+which allow you to return Perl's true and false values, or to return
+an empty list.
+
+If your XSUB will always explicitly return a special SV and won't ever
+require typemap conversions (e.g. it always returns via C<XSRETURN_YES> or
+C<XSRETURN_NO>), then just declare the return type as C<SV*>.
+
+Note that any early return from an XSUB should always be via one of the
+C<XSRETURN> macros and not directly via C<return>; the former will do any
+bookkeeping associated with the argument stack.
+
+=head3 Returning an SV*
+
+More generally, you may want to create and return an SV yourself, rather
+than relying on the boilerplate XSUB code to generate a temporary SV and
+set it to a C-ish value. Here you would declare the return type as C<SV*>.
+For example:
+
+    SV*
+    abc(bool uc)
+      CODE:
+        RETVAL = newSVpv(uc ? "ABC" : "abc", 3);
+      OUTPUT:
+        RETVAL
+
+There is some special processing which happens when using a return type
+such as C<SV*>. First, consider that for a C return type like C<int>, the
+typemap template which sets the temporary SV's value may look something
+like:
+
+    sv_setiv($arg, (IV)$var);
+
+which after expansion may look like:
+
+    sv_setiv(RETVALSV, (IV)RETVAL);
+
+where the temporary SV has previously been assigned to C<RETVALSV>.
+
+Now, if you declare an XSUB with a return type of C<SV*>, you I<might>
+expect the typemap template to look something like:
+
+    sv_setsv($arg, (SV*)$var);
+
+This Perl library function copies the value of one SV to another (the
+XS user's equivalent of the Perl C<$a = $b>).
+
+However, the design decision was made that for the C<SV*> type in
+particular, the typemap template would be
+
+    $arg = $var;
+
+Here is where the special processing comes in. The XS compiler, in the
+case of an output template beginning C<$arg = ...>, skips creating a
+temporary SV, and just returns the SV in C<RETVAL> directly. So the
+typemap template would be expanded to
+
+    ST(0) = RETVAL;
+
+This is faster than copying.
+
+But in addition, for I<any> C<$arg = ...> template (not just the template
+for C<SV*>), the XS compiler makes one further assumption: that the
+expression to the right of the assign evaluates to an SV with a reference
+count count I<one too high>, and so in addition, the XS compiler emits:
+
+    sv_2mortal(RETVAL);
+
+or similar, which causes the reference count of the SV to be decremented
+by one at (typically) the start of the next statement. This makes sense if
+the SV is newly created with one of the C<newSVfoo()> family of functions:
+see the discussion on this in L<An SV's reference count>
+
+However, if the SV comes from elsewhere, for example via a Perl array
+lookup, then its reference count doesn't need to be adjusted, and so the
+mortalising will cause it to be prematurely freed. In this case, you need
+to artificially increase the SV's reference count.
```

Comment:
Could be nice to say "...artificially increase the SV's reference count using SvREFCNT_inc as seen blow"